src/lua_vm/lua_vm.cc

```diff
@@ -0,0 +1,678 @@
+/*
+ * Lua Bindings for RADOS Object Class
+ */
+#include <errno.h>
+#include <setjmp.h>
+#include <string>
+#include <sstream>
+#include <lua.hpp>
+#include "include/types.h"
+#include "json_spirit/json_spirit.h"
+#include "lua_vm/cls_lua_ops.h"
+#include "lua_vm.h"
+
+cls_handle_t h_class;
+cls_method_handle_t h_eval_json;
+cls_method_handle_t h_eval_bufferlist;
+
+/*
+ * Jump point for recovering from Lua panic.
+ */
+static jmp_buf cls_lua_panic_jump;
+
+/*
+ * Handle Lua panic.
+ */
+static int cls_lua_atpanic(lua_State *lua)
+{
+  CLS_ERR("error: Lua panic: %s", lua_tostring(lua, -1));
+  longjmp(cls_lua_panic_jump, 1);
+  return 0;
+}
+
+struct clslua_err {
+  bool error;
+  int ret;
+};
+
+struct clslua_hctx {
+  struct clslua_err error;
+  InputEncoding in_enc;
+  int ret;
+
+  cls_method_context_t *hctx;
```

Comment:
You're right - this code _**is**_ still tuned/specific to objclass but using this commit I can get Mantle to work. Right now, Mantle just uses  bufferlists to exchange metrics between Lua and the MDBalancer. Eventually we'll want to use a struct or something more sophisticated to exchange metrics but this commit is what I need, at a minimum, to get the MDS to load your Lua VM.

The core Lua functions I needed (e.g., `clslua_log()`, `clslua_pcall()`) are in lua_vm.cc, [at line 106] (https://github.com/ceph/ceph/pull/7450/files#diff-64d1a6b29165f2b4a869b0a659c42d63R106). The MDS can't access OSD data, xattr, omap, or env functions, so I put those definitions cls_lua.cc, [at line 365](https://github.com/ceph/ceph/pull/7450/files#diff-6697e0ab00d497e75184b02730dafa19R365).